leshan-core/src/main/java/org/eclipse/leshan/core/request/execute/Arguments.java

```diff
@@ -0,0 +1,260 @@
+        if (argumentList == null) {
+            if (other.argumentList != null)
+                return false;
+        } else if (!argumentList.equals(other.argumentList))
+            return false;
+        return true;
+    }
+
+    /**
+     * Returns number of arguments.
+     */
+    public int size() {
+        return argumentList.size();
+    }
+
+    /**
+     * Check if Arguments has any elements.
+     */
+    public boolean isEmpty() {
+        return size() == 0;
+    }
+
+    /**
+     * Parse String into {@link Arguments}
+     *
+     * <pre>
+     * {@code
+     * arglist = arg *( "," arg )
+     * arg = DIGIT / DIGIT "=" "'" *CHAR "'"
+     * DIGIT = "0" / "1" / "2" / "3" / "4" / "5" / "6" / "7" / "8" / "9"
+     * CHAR = "!" / %x23-26 / %x28-5B / %x5D-7E
+     * }
+     * </pre>
+     * 
+     * @param arglist String contains arguments to parse.
+     * @return {@link Arguments} represents list of arguments.
+     * @throws InvalidArgumentException if text has invalid format.
+     */
+    static Arguments parse(String arglist) throws InvalidArgumentException {
+        List<Argument> argumentList = new ArrayList<>();
+
+        if (arglist != null && !arglist.isEmpty()) {
+            int beginProbe = 0;
+            int validBegin = 0;
+            while (true) {
+                int separatorIndex = arglist.indexOf(',', beginProbe);
+                if (separatorIndex == -1) {
+                    break;
+                }
+                String argument = arglist.substring(validBegin, separatorIndex);
+
+                ArgumentParser argumentParser = new ArgumentParser(argument, arglist);
+                if (argumentParser.isValid()) {
+                    argumentList.add(argumentParser.parse());
+                    validBegin = separatorIndex + 1;
+                }
+                beginProbe = separatorIndex + 1;
+            }
+
+            String argument = arglist.substring(beginProbe);
+            ArgumentParser argumentParser = new ArgumentParser(argument, arglist);
+            argumentList.add(argumentParser.parse());
+        }
+
+        return new Arguments(argumentList);
+    }
+
+    private static class ArgumentParser {
+        private final String digitPart;
+        private final String valueDecorated;
+        private final String content;
+
+        public ArgumentParser(String argument, String content) {
+            this.content = content;
+            String[] keyValue = argument.split("=", 2);
+            digitPart = keyValue[0];
+            valueDecorated = keyValue.length == 1 ? null : keyValue[1];
+        }
+
+        private void validate() throws InvalidArgumentException {
+            if (!isValid()) {
+                throw new InvalidArgumentException("Unable to parse Arguments [%s]", content);
+            }
+        }
+
+        private Argument parse() throws InvalidArgumentException {
+            validate();
+            int digit = Integer.parseInt(digitPart);
```

Comment:
If digitPart is not a number we should raide an InvalidArgumentException

(maybe we can even create an Argument constructor which take a String as Digit and so the validation + raise exception is in the constructor ?)

Reply:
Oh, I missed NumberFormatException :anguished:

The second part: If the only place of Argument creation is while parsing I think it's not a good idea to introduce String-defined digit in costructor to reduce possibilities of producing error.

For sure I should improve validation while parsing.